pyteal/ast/router.py

```diff
@@ -0,0 +1,578 @@
+from dataclasses import dataclass
+from typing import Any, cast, Optional
+
+import algosdk.abi as sdk_abi
+
+from pyteal.config import METHOD_ARG_NUM_LIMIT
+from pyteal.errors import TealInputError
+from pyteal.types import TealType
+
+from pyteal.ast import abi
+from pyteal.ast.subroutine import (
+    OutputKwArgInfo,
+    SubroutineFnWrapper,
+    ABIReturnSubroutine,
+)
+from pyteal.ast.cond import Cond
+from pyteal.ast.expr import Expr
+from pyteal.ast.app import OnComplete, EnumInt
+from pyteal.ast.int import Int
+from pyteal.ast.seq import Seq
+from pyteal.ast.methodsig import MethodSignature
+from pyteal.ast.naryexpr import And, Or
+from pyteal.ast.txn import Txn
+from pyteal.ast.return_ import Approve
+
+"""
+Notes:
+- On a BareApp Call, check
+  - [x] txn NumAppArgs == 0
+  - [x] On-Completion should match (can be a list of On-Completion here)
+  - [x] Must execute actions required to invoke the method
+
+- On Method Call, check
+  - [x] txna ApplicationArgs 0 == method "method-signature"
+  - [x] On-Completion should match (only one On-Completion specified here?)
+  - [x] non void method call should log with 0x151f7c75 return-method-specifier
+        (kinda done in another PR to ABI-Type)
+  - [x] redirect the method arguments and pass them to handler function
+        (kinda done, but need to do with extraction and (en/de)-code)
+  - [x] Must execute actions required to invoke the method
+  - [x] extract arguments if needed
+        (decode txna ApplicationArgs 15 if there exists, and extract arguments to feed method)
+
+Notes for OC:
+- creation conflict with closeout and clearstate
+- must check: txn ApplicationId == 0 for creation
+- clearstate AST build should be separated with other OC AST build
+"""
+
+
+@dataclass
+class ProgramNode:
+    """
+    This class contains a condition branch in program AST, with
+    - `condition`: logical condition of entering such AST branch
+    - `branch`: steps to execute the branch after entering
+    - `method_info` (optional): only needed in approval program node, constructed from
+        - SDK's method
+        - ABIReturnSubroutine's method signature
+    """
+
+    condition: Expr
+    branch: Expr
+    method_info: Optional[sdk_abi.Method]
+    ast_order_indicator: "ConflictMapElem"
+
+
+ProgramNode.__module__ = "pyteal"
+
+
+@dataclass
+class ConflictMapElem:
```

Comment:
First, my apologies for no tests and no docstrings to explain what is happening in this class.

There are 2 things I am aware of in router should be circumvented:
- re-registering a method under same OC: it is not good to re-registering a same method under same OC, for both bare-app-call and method-call cases.
  - Will the resulting program hurt from it, probably not.
  - But there will be a dead branch one never enter, for the same condition would trigger a same earlier branch.
- overshadowing in AST under same OC:
  - consider: first registering a bare-call `Approve()` for `OC=NoOp`, then register method `add` in `OC=NoOp`.
  - ignoring my fault in generating the condition (`Txn.num_application_arg == 0`)
  - then the first registered `approve` bare call would overshadow the later registered `add` method, since call with `NoOp` will go into `Approve()` first.

One way to resolve the problem of overshadowing is that, sort by following principle:
- method call comes first in AST, then bare call in AST
- on_creation comes first in AST, then !on_creation

This would ensure that, assuming there is at most one barecall in each OC, then the method calls and bare calls are never overshadowed.

Finally, another restriction comes from previous argument, which is for each OC, there is at most one bare call registering.

Now I felt like there are a lot of policy need to be taken care of, and I might be missing something, or we should divert to another path, which is not guaranteeing any check, and use at own cautions about generated program...